src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+    case message_hash        :
+    case no_shake            :
+        return 1;
+    }
+
+    /* abide compilers */
+    return 0;
+}
+
+
+static int dtls13_get_rn_mask(WOLFSSL *ssl, byte *ciphertext, byte *mask,
+                              word16 side)
+{
+    byte *key;
+    Aes aes;
+    int ret;
+
+    /* FIXME: implement other ciphers */
+    if (ssl->specs.bulk_cipher_algorithm != wolfssl_aes_gcm)
+        return WOLFSSL_NOT_IMPLEMENTED;
+
+    ret = wc_AesInit(&aes, ssl->heap, -1);
+    if (ret != 0)
+        return ret;
+
+    key = (side == WOLFSSL_SERVER_END) ?
+        ssl->keys.server_sn_key : ssl->keys.client_sn_key;
+
+    ret = wc_AesSetKey(&aes, key, ssl->specs.key_size, NULL, AES_ENCRYPTION);
+    if (ret != 0)
+        goto exit;
+
+    ret = wc_AesEncryptDirect(&aes, mask, ciphertext);
+    if (ret != 0)
+        goto exit;
+
+ exit:
+    wc_AesFree(&aes);
+    return ret;
+}
+
+static int dtls13_encrypt_decrypt_record_number(WOLFSSL *ssl,
+                                                byte *seq,
+                                                int seq_length,
+                                                byte *ciphertext,
+                                                enum rn_direction dir)
+
+{
+    byte mask[RN_MASK_SIZE];
+    int ret, i;
+    word8 side;
+
+    if (dir == PROTECT)
+        side = ssl->options.side;
+    else
+        side = (ssl->options.side == WOLFSSL_SERVER_END) ?
+            WOLFSSL_CLIENT_END : WOLFSSL_SERVER_END;
+
+    ret = dtls13_get_rn_mask(ssl, ciphertext, mask, side);
+    if (ret != 0)
+        return ret;
+
+    for (i = 0; i < seq_length; ++i)
+        seq[i] ^= mask[i];
+
+    return 0;
+}
+
+/* this is a clone of DtlsMsgDrain */
+static int dtls13_process_buffered_messages(WOLFSSL *ssl)
+{
+    DtlsMsg *msg = ssl->dtls_rx_msg_list;
+    word32 idx = 0;
+    int ret = 0;
+
+    WOLFSSL_ENTER("dtls13_process_buffered_messages()");
+    while (msg != NULL) {
+
+        /* message not in order */
+        if (ssl->keys.dtls_expected_peer_handshake_number != msg->seq)
+            break;
+
+        /* message not complete */
+        if (msg->fragSz != msg->sz)
+            break;
+
+        ret = DoTls13HandShakeMsgType(ssl, msg->msg, &idx, msg->type,
+                                      msg->sz, msg->sz);
+        if (ret != 0)
+            break;
+
+        /* TODO: update the retransmission logic with new message processed */
+
+        /* FIXME: move in a better place. handle overflow? */
+        ssl->keys.dtls_expected_peer_handshake_number++;
+
+        ssl->dtls_rx_msg_list = msg->next;
+        DtlsMsgDelete(msg, ssl->heap);
+        msg = ssl->dtls_rx_msg_list;
+        ssl->dtls_rx_msg_list_sz--;
+    }
+
+    WOLFSSL_LEAVE("dtls13_process_buffered_messages()", ret);
+
+    return ret;
+}
+
+static int dtls13_next_message_complete(WOLFSSL *ssl) {
+    return ssl->dtls_rx_msg_list != NULL &&
+        ssl->dtls_rx_msg_list->fragSz >= ssl->dtls_rx_msg_list->sz;
```

Comment:
Wouldn't `ssl->dtls_rx_msg_list->fragSz > ssl->dtls_rx_msg_list->sz` be an error? Is there a case where this is possible?